fastjsonschema/generator.py

```diff
@@ -210,75 +332,92 @@ def l(self, line, *args, **kwds):
             with self.l('if {variable} not in {enum}:'):
                 self.l('raise JsonSchemaException("Wrong!")')
         """
-        spaces = ' ' * self.INDENT * self._indent
-
         name = self._variable_name
-        if name and '{' in name:
+        if expand_locals and '{name}' in line and name and '{' in name:
             name = '"+"{}".format(**locals())+"'.format(self._variable_name)
 
         context = dict(
             self._definition or {},
             variable=self._variable,
             name=name,
-            **kwds
+            **kwds,
         )
-        line = line.format(*args, **context)
-        line = line.replace('\n', '\\n').replace('\r', '\\r')
-        self._code.append(spaces + line)
-        return line
+        code = line.format(*args, **context)
+        code = code.replace('\n', '\\n').replace('\r', '\\r')
+        self._emit(self._indent, code, deduplicate)
 
-    def e(self, string):
+    @staticmethod
+    def e(string):
         """
         Short-cut of escape. Used for inserting user values into a string message.
 
         .. code-block:: python
-
             self.l('raise JsonSchemaException("Variable: {}")', self.e(variable))
         """
         return str(string).replace('"', '\\"')
 
+    def throw(self, message: str, *args, **kwargs):
+        """
+        Generate code for `raise JsonSchemaException(...)`
+
+        This function has two purposes:
+        - to keep the code less verbose
+        - distinguish between final validation failures and probing validation failures
+
+        :param message: string to be included in the exception
+        """
+        expand = not self.is_probing
+        if self._verbose:
+            message += '\\n  caused by {path} {rule}'
+            kwargs.update(path=self.context_path(), rule=self._definition)
+        self.l('raise JsonSchemaException("{}")'.format(message if expand else ''),
```

Comment:
Is there some real performance gain with this probing? If not, I think it's then uselessly complex.

Reply:
not really, but when the the _definition is huge... it is huge in the compiled schema